### PR TITLE
Linode Longview package and module

### PR DESCRIPTION
**Pull request by rvl**
Longview is a perl script used for sending server metrics to Linode virtual private server hosting.

### CODE REVIEW

File: nixos/modules/services/monitoring/longview.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        default = false;
+        description = ''
+          If enabled, system metrics will be sent to Linode LongView.
+        '';
+      };
+
+      apiKey = mkOption {
+        type = types.str;
+        description = ''
+          Longview API key. To get this, look in Longview settings.
```
Comment (domenkozar) [inline nixos/modules/services/monitoring/longview.nix]: Maybe add the url to the settings?

File: pkgs/servers/monitoring/longview/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  src = fetchFromGitHub {
+    owner = "linode";
+    repo = "longview";
+    rev = "5bcc9b60896b72de2d14f046f911477c26eb70ba";
+    sha256 = "1i6va44bx2zfgbld7znf1slph0iqidlahq2xh3kd8q4lhvbrjn02";
+  };
+
+  patches =
+    [ # Allow the paths of the config files to be set at runtime
+      ./runtime-paths.patch
```
Comment (domenkozar) [inline pkgs/servers/monitoring/longview/default.nix]: would be nice if you submited this patch upstream and added a link to the bug here

File: pkgs/servers/monitoring/longview/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+      # log to systemd journal
+      ./log-stdout.patch
+
+      # configurable pid file location
+      ./pidfile.patch
```
Comment (domenkozar) [inline pkgs/servers/monitoring/longview/default.nix]: same for this patch

File: pkgs/servers/monitoring/longview/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    ];
+
+  buildInputs = [ perl makeWrapper glibc ]
+    ++ (with perlPackages; [ LWPUserAgent LWPProtocolHttps MozillaCA CryptSSLeay IOSocketInet6 LinuxDistribution JSONPP JSON LogLogLite TryTiny DBI DBDmysql ]);
```
Comment (domenkozar) [inline pkgs/servers/monitoring/longview/default.nix]: (nitpick) could you break this into two lines?

File: nixos/modules/services/monitoring/longview.nix [hunk 5]
```diff
@@ -0,0 +1,66 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.longview;
+
+  pidFile = "/run/longview.pid";
+
+  apacheConf = pkgs.writeText "Apache.conf" ''
```
Comment (domenkozar) [inline nixos/modules/services/monitoring/longview.nix]: all these three config files are commented, is this supposed to work?

**Comment by rvl** [comment]
Hello Domen,

Thanks for reviewing. I have reworked the package and module based on your comments and also from new things I have learnt about Nix.

* URL for getting the API key is now in the config help
* WRT to upstreaming patches, I feel this is unlikely to succeed because NixOS is not one of the supported OS for Linode. However I have found better ways to do the configuration which avoids patching the source.
* The commented config files are only there as placeholders so that the script doesn't fail. Fully integrating with those services exceeds my present need/ambition. With the use of `environment.etc`, the user should be able to configure if necessary.


Cheers,

Rodney

**Comment by domenkozar** [comment]
@rvl `environment.etc` are symlinked to `/nix/store` so they are not writable. You have to expose configuration files through `options` to be settable in `configuration.nix`

**Comment by rvl** [comment]
OK thanks @domenkozar. I am still trying to understand how module configuration corresponds to `/etc`. Anyway, these config files are now built from the module options.

**Comment by rickynils** [comment]
@rvl From your last comment, I'm not sure if you still consider this work-in-progress or not?

**Comment by rvl** [comment]
Hello @rickynils @domenkozar , sorry to be vague. I have added commit c28bb1b which addresses the `environment.etc` comment. So this branch is ready for further review/merge.

**Comment by rvl** [comment]
OK I have added a little advice to the config descriptions. I'm not sure what we can do about the general problem of passwords in the nix store.

**Comment by rickynils** [comment]
@rvl, no in general there is no good solution for storing secrets in the nix store. The best thing currently is if the service can read secrets from a run time file path. Then the nix user can choose to either build the secret into the store or to put the secret outside the store and deploy it through some other means than a nix closure. But if your service needs the secrets inside a configuration file, with no possibility to refer to an external file, it gets trickier. You could implement an activation script that generates the config file when the service start, or you can simply ignore the problem and document the fact that people should not use the service if they have secrets they care about.

**Comment by rvl** [comment]
Hi @rickynils, I would happily take part in any effort to patch the (numerous) services which currently need a password in a config file — once a general solution is decided upon.

In the meantime, let's merge this puppy.

I have rebased on latest master to resolve the merge conflict.

File: nixos/modules/services/monitoring/longview.nix [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        type = types.bool;
+        default = false;
+        description = ''
+          If enabled, system metrics will be sent to Linode LongView.
+        '';
+      };
+
+      apiKey = mkOption {
+        type = types.str;
+        example = "01234567-89AB-CDEF-0123456789ABCDEF";
```
Comment (nbp) [inline nixos/modules/services/monitoring/longview.nix]: We should have alternative to avoid having these kind of information public.

**Comment by nbp** [comment]
@rvl, Please add a note in the `rl-unstable.xml` release note file.

**Comment by rvl** [comment]
Release note in #11291

**Comment by rvl** [comment]
@domenkozar @rickynils @nbp Thanks for reviewing and merging.

**Pull request merged**